plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/utils/ProjectImportEnablingScope.java

```diff
@@ -271,16 +306,22 @@ public IEObjectDescription getSingleElement(EObject object) {
 		return result.values();
 	}
 
-	private IN4JSProject findProject(N4JSProjectName projectName, IN4JSProject project) {
+	private IN4JSProject findProject(N4JSProjectName projectName, IN4JSProject project,
+			boolean preferDefinitionProject) {
+
 		if (Objects.equals(project.getProjectName(), projectName)) {
 			return project;
 		}
 
 		Iterable<? extends IN4JSProject> dependencies = project.getSortedDependencies();
-		for (IN4JSProject p : dependencies) {
-			if (Objects.equals(p.getDefinesPackageName(), projectName)) {
-				return p;
+		if (preferDefinitionProject) {
+			for (IN4JSProject p : dependencies) {
+				if (Objects.equals(p.getDefinesPackageName(), projectName)) {
+					return p;
+				}
 			}
+		}
+		for (IN4JSProject p : dependencies) {
```

Comment:
Check whether dependency sort order already suffices here to avoid a second loop over all dependencies.